### PR TITLE
[js/webgpu] Support f16 uniform

### PR DESCRIPTION
**Pull request by axinging**

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @xhcao @hujiajie @gyagp PTAL

**Review by guschmue — APPROVED** [review]

File: js/web/lib/wasm/jsep/webgpu/ops/common.ts [hunk 1]
```diff
@@ -330,18 +330,25 @@ export const sumVector = (name: string, components: number) => {
...
-      return `${name}[${Math.floor(index / 4)}][${index % 4}]`;
-    }
-  } else {
-    return length > 1 ? `${name}[${index}]` : name;
-  }
-};
+export const getElementAt =
+    (name: string, index: number|string, length: number, type?: UniformDataElementType): string => {
+      if (name.startsWith('uniforms.') && length > 4) {
+        if (typeof (index) === 'string') {
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/common.ts:338]: if index is string, and the type is `f16`, this seems generate wrong result.

Comment (axinging) [reply js/web/lib/wasm/jsep/webgpu/ops/common.ts:338]: done

**Review by guschmue — APPROVED** [review]

**Review by fs-eire — APPROVED** [review]

File: js/web/lib/wasm/jsep/backend-webgpu.ts [hunk 2]
```diff
@@ -461,6 +474,9 @@ export class WebGpuBackend {
           new Int32Array(arrayBuffer, offset, data.length).set(data);
         } else if (v.type === 'uint32') {
           new Uint32Array(arrayBuffer, offset, data.length).set(data);
+        } else if (v.type === 'float16') {
+          // TODO: use Float16Array.
+          new Uint16Array(arrayBuffer, offset, data.length).set(data);
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/backend-webgpu.ts:479]: Do we need to do some preprocessing to pass cpu data to Uint16Array if the cpu data is treated as float32? For example, the pad2~pad10, the `constant_value` is float not `T` as input.
cc  @jzm-intel

File: onnxruntime/core/providers/js/operators/pad.cc [hunk 3]
```diff
@@ -24,7 +24,7 @@ ONNX_OPERATOR_VERSIONED_KERNEL_EX(
     12,
     kJsExecutionProvider,
     (*KernelDefBuilder::Create())
-        .TypeConstraint("T", DataTypeImpl::GetTensorType<float>())
+        .TypeConstraint("T", JsepSupportedFloatTypes())
```
Comment (qjia7) [inline onnxruntime/core/providers/js/operators/pad.cc:27]: For version <=10, we can use `JsepSupportedFloatTypes()`. After versions (>10), use `JsepSupportedDataTypes()`?

**Review by qjia7 — COMMENTED** [review]
Sorry for the late response. Two more comments to supplement. You can do it in follow-up PRs.

File: js/web/lib/wasm/jsep/backend-webgpu.ts (hunk 2 above)
Comment (qjia7) [reply js/web/lib/wasm/jsep/backend-webgpu.ts:479]: Sorry, my example is not correct for this case. If `constant_value` is float, the code will directly go to the `float32` path not `float16` case.  But the concern is still there.

Comment (axinging) [reply js/web/lib/wasm/jsep/backend-webgpu.ts:479]: Yes, user should make sure the "data" is stored float16 bits,  for example: If "data" is 1.0, then the corresponding float16 bits is 0x3c00. I am trying to add some warning msg to inform user that this should be unit test only: https://github.com/microsoft/onnxruntime/pull/19307/files#diff-006fc84d3997f96a29b8033bd2075d6a0a9509211bd5812a6b934fc74fedfd9dR520

**Pull request merged**